ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/PreJoinProjectNode.java

```diff
@@ -129,6 +141,32 @@ protected Set<ColumnReferenceExp> validateColumns(final RequiredColumns required
     return super.validateColumns(builder.build());
   }
 
+  @Override
+  public SchemaKStream<?> buildStream(final PlanBuildContext buildContext) {
+    final SchemaKStream<?> stream = getSource().buildStream(buildContext);
+
+    final List<ColumnName> keyColumnNames = getSchema().key().stream()
+        .map(Column::name)
+        .collect(Collectors.toList());
+    if (stream instanceof SchemaKTable || !isSelfJoin) {
+      return stream.select(
+          keyColumnNames,
+          getSelectExpressions(),
+          buildContext.buildNodeContext(getId().toString()),
+          buildContext,
+          getFormatInfo()
+      );
+    } else {
+      return stream.noOpSelect(
```

Comment:
I know this PR is still a draft, but it'd be good to make sure we have lots of comments and documentation around why we're doing this and what the point of a `noOpSelect` is. I'm sure future people will be very confused!